### PR TITLE
Closes Issue/96

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

File: src/directory/managed_directory.rs [hunk 1]
```diff
@@ -65,53 +94,98 @@ impl ManagedDirectory {
     /// an error is simply logged, and the file remains in the list of managed
     /// files.
     pub fn garbage_collect(&mut self, living_files: HashSet<PathBuf>) {
-        let mut managed_has_changed: bool = false;
-        {
-            let mut files_to_delete = vec!();
-            let mut managed_paths_write = self.managed_paths.write().unwrap();
+        let mut files_to_delete = vec!();
+        {   // releasing the lock as .delete() will use it too.
+            let mut meta_informations_wlock = self.meta_informations.write().unwrap();
```
Comment (KodrAus) [inline src/directory/managed_directory.rs]: nit: We could use `expect` rather than `unwrap`.

**Review by KodrAus — APPROVED** [review]
This is a whole lotta code I don't think I could reasonably review. It's a lot more complexity, but if it tames those merges then I think it's worth it.

File: src/core/segment.rs [hunk 2]
```diff
@@ -43,6 +43,10 @@ impl Segment {
         self.index.schema()
     }
 
+    pub fn index(&self,) -> &Index {
```
Comment (fulmicoton) [inline src/core/segment.rs]: remove as it is now uselss

File: src/core/segment.rs [hunk 3]
```diff
@@ -66,6 +70,11 @@ impl Segment {
         self.meta.relative_path(component)
     }
 
+    pub fn protect_from_delete(&self, component: SegmentComponent) -> FileProtection {
```
Comment (fulmicoton) [inline src/core/segment.rs]: comment and test

File: src/datastruct/stacker/heap.rs [hunk 4]
```diff
@@ -109,11 +110,7 @@ struct InnerHeap {
 /// We use this unsafe trick to make unit test
 /// way faster.
 fn allocate_fast(num_bytes: usize) -> Vec<u8> {
```
Comment (fulmicoton) [inline src/datastruct/stacker/heap.rs]: remove function

File: src/directory/error.rs [hunk 5]
```diff
@@ -27,7 +27,7 @@ impl From<io::Error> for OpenWriteError {
     }
 }
 
-/// Error that may occur when accessing a file (read, or delete)
+/// Error that may occur when accessing a file read
 #[derive(Debug)]
 pub enum FileError {
```
Comment (fulmicoton) [inline src/directory/error.rs]: Rename ReadError

File: src/directory/managed_directory.rs (hunk 1 above)
Comment (fulmicoton) [inline src/directory/managed_directory.rs]: expect(...)

double check that expect is ok here.
ie: the lock is released right away, with code that cannot panick (no io, no call to other library etc.)

File: src/directory/managed_directory.rs [hunk 6]
```diff
@@ -65,53 +94,98 @@ impl ManagedDirectory {
...
             }
         }
-        if managed_has_changed {
+
+
+        if !deleted_files.is_empty() {
+            // update the list of managed files by removing 
+            // the file that were removed.
+            {
+                let mut meta_informations_wlock = self.meta_informations.write().unwrap();
```
Comment (fulmicoton) [inline src/directory/managed_directory.rs]: unwrap()

File: src/indexer/delete_queue.rs [hunk 7]
```diff
@@ -1,74 +1,274 @@
...
+        else {
+            false
+        }
     }
 
-    pub fn clear(&self) {
-        self.0.write().unwrap().clear();
+    /// Get the current delete operation.
+    /// Calling `.get` does not advance the cursor.
+    pub fn get<'a>(&'a mut self) -> Option<&'a DeleteOperation> {
```
Comment (fulmicoton) [inline src/indexer/delete_queue.rs]: lifetime useless

File: src/indexer/delete_queue.rs [hunk 8]
```diff
@@ -1,74 +1,274 @@
...
+//   will include all future delete operation (and no past operations).
+// - cloning an existing cursor returns a new cursor, that
+//   is at the exact same position, and can now advance independantly 
+//   from the original cursor.
 #[derive(Default)]
 struct InnerDeleteQueue {
-    ro_chunks: DeleteQueueSnapshot,
-    last_chunk: Vec<DeleteOperation>,
+    writer: Vec<DeleteOperation>,
+    last_block: Option<Arc<Block>>, // TODO last block... is that ok.
```
Comment (fulmicoton) [inline src/indexer/delete_queue.rs]: remove comment

File: src/indexer/delete_queue.rs [hunk 9]
```diff
@@ -1,74 +1,274 @@
...
+}
+
+
+impl DeleteQueue {
+
+    // Creates a new delete queue.
+    pub fn new() -> DeleteQueue {
+        
+        let delete_queue = DeleteQueue {
+            inner: Arc::new(RwLock::new(InnerDeleteQueue::default()))
```
Comment (fulmicoton) [inline src/indexer/delete_queue.rs]: Arc::default

File: src/indexer/delete_queue.rs [hunk 10]
```diff
@@ -1,74 +1,274 @@
...
+        };
+            
+        let next_block = NextBlock::from(delete_queue.clone());
+        {
+            let mut delete_queue_wlock = delete_queue.inner.write().unwrap();
```
Comment (fulmicoton) [inline src/indexer/delete_queue.rs:44]: .expect()

File: src/indexer/delete_queue.rs [hunk 11]
```diff
@@ -1,74 +1,274 @@
...
-            let new_ro_chunk = mem::replace(&mut self.last_chunk, new_operations);
-            self.ro_chunks.push(new_ro_chunk)
+    // Creates a new cursor that makes it possible to 
+    // consume future delete operations.
+    // 
+    // Past delete operations are not accessible.
+    pub fn cursor(&self) -> DeleteCursor {
+        let last_block = self.inner
+            .read()
+            .unwrap()
```
Comment (fulmicoton) [inline src/indexer/delete_queue.rs]: expect

File: src/directory/managed_directory.rs [hunk 12]
```diff
@@ -65,53 +94,98 @@ impl ManagedDirectory {
...
+                }
+            }
             if let Err(_) = self.save_managed_paths() {
                 error!("Failed to save the list of managed files.");
             }
         }
+
+    }
+
+    pub fn protect_file_from_delete(&self, path: &Path) -> FileProtection {
```
Comment (fulmicoton) [inline src/directory/managed_directory.rs]: add unit test for this

File: src/indexer/merge_policy.rs [hunk 13]
```diff
@@ -43,3 +43,33 @@ impl MergePolicy for NoMergePolicy {
...
+
+#[cfg(test)]
+pub mod tests {
+
+    use super::*;
+    use core::SegmentId;
+    use core::SegmentMeta;
+
+    #[derive(Debug)]
+    pub struct MergeWheneverPossible;
```
Comment (fulmicoton) [inline src/indexer/merge_policy.rs]: add comment

File: src/indexer/merger.rs [hunk 14]
```diff
@@ -34,7 +34,11 @@ struct DeltaPositionComputer {
 
 impl DeltaPositionComputer {
     fn new() -> DeltaPositionComputer {
-        DeltaPositionComputer { buffer: iter::repeat(0u32).take(512).collect::<Vec<u32>>() }
+        DeltaPositionComputer { 
+            buffer: iter::repeat(0u32)
```
Comment (fulmicoton) [inline src/indexer/merger.rs]: use common::allocate

File: src/indexer/segment_entry.rs [hunk 15]
```diff
@@ -22,35 +24,44 @@ impl SegmentState {
 pub struct SegmentEntry {
```
Comment (fulmicoton) [inline src/indexer/segment_entry.rs:38]: doc

File: src/indexer/segment_entry.rs [hunk 16]
```diff
@@ -22,35 +24,44 @@ impl SegmentState {
...
             state: SegmentState::Ready,
-            doc_to_opstamp: DocToOpstampMapping::None,
+            delete_bitset: delete_bitset,
+            delete_cursor: delete_cursor,
         }
     }
 
-    pub fn doc_to_opstamp(&self) -> &DocToOpstampMapping {
-        &self.doc_to_opstamp
+    pub fn delete_bitset(&self,) -> Option<&BitSet> {
```
Comment (fulmicoton) [inline src/indexer/segment_entry.rs]: doc

File: src/indexer/segment_entry.rs [hunk 17]
```diff
@@ -22,35 +24,44 @@ impl SegmentState {
...
+        self.delete_bitset.as_ref()
+    }
+
+    pub fn set_meta(&mut self, segment_meta: SegmentMeta) {
```
Comment (fulmicoton) [inline src/indexer/segment_entry.rs]: doc

File: src/indexer/segment_entry.rs [hunk 18]
```diff
@@ -22,35 +24,44 @@ impl SegmentState {
...
+        self.meta = segment_meta;
+    }
+
+    pub fn delete_cursor(&mut self) -> &mut DeleteCursor {
```
Comment (fulmicoton) [inline src/indexer/segment_entry.rs]: doc

File: src/indexer/segment_manager.rs [hunk 19]
```diff
@@ -71,23 +66,29 @@ impl SegmentManager {
         segment_entries
     }
 
+    pub fn num_segments(&self,) -> usize {
```
Comment (fulmicoton) [inline src/indexer/segment_manager.rs]: is it used? if not remove

File: src/indexer/segment_manager.rs [hunk 20]
```diff
@@ -118,31 +119,19 @@ impl SegmentManager {
...
-                    let segment_id = segment_meta.id();
-                    let mut segment_entry = SegmentEntry::new(segment_meta);
-                    if let Some(state) = self.segment_state(&segment_id) {
-                        segment_entry.set_state(state);
-                    }
-                    segment_entry
-                })
-                .collect::<Vec<_>>();
+    pub fn commit(&self, mut segment_entries: Vec<SegmentEntry>) {
+        // TODO is still relevant!?
```
Comment (fulmicoton) [inline src/indexer/segment_manager.rs]: address comment

**Pull request merged**